### PR TITLE
Deselect Chapters - Sync

### PR DESCRIPTION
**Pull request by ashiagr**
Part of: #1807 
## Description
**Blocked by backend changes**

Adds sync to the chapter deselection (`for podcast episodes only`).

Podcast with chapters you can use to test:
* Upgrade: https://pca.st/upgrade

> [!WARNING]
> Do not merge this PR until Sync API changes are merged

## To test

Prerequisite:
- Local Sync API setup
- Checkout Sync API project branch `feature/deselect-chapters-save-update`
- Apply [add_deselected_chapters_to_legacy_sync_record.patch](https://github.com/Automattic/pocket-casts-android/files/14534318/add_deselected_chapters_to_legacy_sync_record.patch) to Sync API project (I'll sync with Leandro to apply these changes)
- Point `SERVER_API_URL` in `pocket-casts-android` `base.gradle` to the local Sync API server
- Paid account

### Test.1 Migration

Since we added a new field `deselected_chapters_modified` to db, test that migration works properly

1. Checkout commit where the field is not added (e.g. 18872f23 )  
2. Install the app
3. Checkout this PR
4. Install and launch the app
5. Migration should be successful

### Test.2 Basic Sync

1. Login to a paid account 
6. Play an episode with chapters
7. Go to Chapters, tap "Skip Chapters"
8. Deselect a few chapters
9. Tap Done
10. Dismiss the player
11. Go to Profile > Refresh Now 
12. Delete the app
13. Reinstall
14. Login to the same account
15. Play the episode that was playing (needed so chapters are displayed)
16. ✅ Your previously deselected chapters should be deselected

### Test.3 Multi-device sync

D1 is device 1, D2 is device 2.

1. Login to the same paid account on both devices
2. D1: Play an episode with chapters in
3. Go to Chapters, tap "Skip Chapters"
4. Deselect a few chapters
5. Tap Done
6. Dismiss the player
7. Go to Profile > Refresh Now 
8. On D2, go to Profile > Refresh Now (or pull to refresh)
9. Open the Player and go to Chapters (you might need to play the episode so chapters appear)
10. ✅ Your deselected chapters on D1 should reflect on D2

You can play with this by changing chapters in one and checking if it's reflected on the other after updating.

### Test.4 Syncing episode not in Up Next

D1 is device 1, D2 is device 2.

1. Login to the same paid account on both devices
2. D1: Play an episode with chapters in
3. Go to Chapters, tap "Skip Chapters"
4. Deselect a few chapters
5. Tap Done
6. Dismiss the player
8. Tap and hold on the miniplayer and tap "Close and clear Up Next"
8. Go to Profile > Refresh Now 
9. On D2, go to Profile > Refresh Now (or pull to refresh)
10. Go to Listening History and find the episode that was playing
11. Play it
12. ✅ Your deselected chapters on D1 should reflect on D2

*** Important: 
1. Repeat `Test.2`, `Test.3` and `Test.4` after disabling `Settings Sync` feature flag in `Profile` -> `Settings` -> `Beta features`.
2. Smoke test that the app doesn't crash on sync after pointing `SERVER_API_URL` to https://api.pocketcasts.net where the changes are not deployed yet.

## Checklist
- [ ] If this is a user-facing change, I have added an entry in CHANGELOG.md
- [x] Ensure the linter passes (`./gradlew spotlessApply` to automatically apply formatting/linting)
- [ ] I have considered whether it makes sense to add tests for my changes
- [ ] All strings that need to be localized are in `modules/services/localization/src/main/res/values/strings.xml`
- [ ] Any jetpack compose components I added or changed are covered by compose previews
 
#### I have tested any UI changes...

- [ ] with different themes
- [ ] with a landscape orientation
- [ ] with the device set to have a large display and font size
- [ ] for accessibility with TalkBack

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>ChapterIndices</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/UpNextQueue.kt
Comment (MiSikora) [inline modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/UpNextQueue.kt:140]: Is this check important? I'm genuinely asking, not suggesting. It feels like it shouldn't matter if we update the list with another list if the contain the same elements but in different order.

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/PlaybackManager.kt
Comment (MiSikora) [inline modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/PlaybackManager.kt:964]: Is it necessary to switch to the `Main` dispatcher here? If it is `blockingFirst()` will block the main thread. It can be avoided with `playbackStateRelay.firstOrError().await()`.

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/PlaybackManager.kt
Comment (MiSikora) [inline modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/PlaybackManager.kt:985]: Similarly here. While there is no thread defined here, the usage of this method in the `MainActivity` flow will block the main thread.

https://github.com/Automattic/pocket-casts-android/blob/6c139b2573a64d6c62cb35e3c2fbad65e35a140b/app/src/main/java/au/com/shiftyjelly/pocketcasts/ui/MainActivity.kt#L763

**Review by MiSikora — COMMENTED** [review]
I haven't tested yet because it will take me some time to setup the server but I checked the code and left some comments.

File: modules/services/model/src/main/java/au/com/shiftyjelly/pocketcasts/models/entity/BaseEpisode.kt
Comment (ashiagr) [reply modules/services/model/src/main/java/au/com/shiftyjelly/pocketcasts/models/entity/BaseEpisode.kt]: Converted in https://github.com/Automattic/pocket-casts-android/pull/1916/commits/ec96fee5ff70b96089b7d1ec2eefcac112483396

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/UpNextQueue.kt
Comment (ashiagr) [reply modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/UpNextQueue.kt]: Ah! I was comparing strings in an earlier attempt when sorting was required. With lists, it isn't. Removed in https://github.com/Automattic/pocket-casts-android/pull/1916/commits/615b5fb409089608706ededf170642b299863664

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/PlaybackManager.kt
Comment (ashiagr) [reply modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/PlaybackManager.kt]: I noticed in one instance that toggling was reverted without switching to the main dispatcher. I can't reproduce it now. I have removed the switch but I'll continue testing it.

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/PlaybackManager.kt
Comment (ashiagr) [reply modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/PlaybackManager.kt]: I wrapped it in `launch{...}` in https://github.com/Automattic/pocket-casts-android/pull/1916/commits/c75e76036e69db6f167119e11cf3c809801ed2db

**Review by MiSikora — COMMENTED** [review]
Checked with local sync API and everything worked. I'll revisit with the approval once server changes gets merged to staging.

**Comment by ashiagr** [comment]
@MiSikora, server changes are merged to staging. Can you revisit approving this PR?

**Review by MiSikora — APPROVED** [review]
Sure, I though of waiting for your follow-up server changes to get merged but it's not necessary. LGTM!

**Comment by ashiagr** [comment]
Thank you!

> I though of waiting for your follow-up server changes to get merged but it's not necessary. 

Those got merged as well: https://github.com/Automattic/pocket-casts-sync-api/pull/947

**Pull request merged**
